os/kernel/sched/sched_addDeadtask.c

```diff
@@ -0,0 +1,281 @@
+};
+
+typedef struct dead_name_s dead_name_t;
+
+/* list to store dead pid node */
+sq_queue_t dead_pidList;
+
+/* list to store dead pid's name node. Note that, there can be 
+ * multiple pid with same name. So to minimize memory usage,
+ * we are using another list for name */
+sq_queue_t dead_nameList;
+
+/****************************************************************************
+ * Global Variables
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Variables
+ ****************************************************************************/
+
+/****************************************************************************
+ * Private Function Prototypes
+ ****************************************************************************/
+
+dead_pid_t *allocate_pid_node(int pid)
+{
+    dead_pid_t *node = (dead_pid_t *)malloc(sizeof(dead_pid_t));
+    if (!node) {
+        dbg("ERROR: memory allocation error\n");
+        return NULL;
+    }
+    node->pid = pid;
+    return node;
+}
+dead_name_t *allocate_name_node(char *name)
+{
+    dead_name_t *node = (dead_name_t *)malloc(sizeof(dead_name_t));
+    if (!node) {
+        dbg("ERROR: memory allocation error\n");
+        return NULL;
+    }
+    strncpy(node->name, name, CONFIG_TASK_NAME_SIZE);
+    node->ref_cnt = 0;
+    return node;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: sched_deadtasklistinit
+ *
+ * Description:
+ *   This function initializes dead task info list
+ *
+ * Inputs:
+ *   None
+ *
+ ****************************************************************************/
+
+void sched_deadtasklistinit()
+{
+    sq_init(&dead_pidList);
+    sq_init(&dead_nameList);
+}
+
+/****************************************************************************
+ * Name: sched_addDeadtask
+ *
+ * Description:
+ *   This function adds a task's pid and name info to the list.
+ *   This should be called before deleting a task.
+ *
+ * Inputs:
+ *   pid: pid of the deleting task
+ *   name: name of the task
+ *
+ ****************************************************************************/
+
+void sched_addDeadtask(int pid, char *name)
+{
+    dead_pid_t *pid_head = (dead_pid_t *)dead_pidList.head;
+    dead_pid_t *pid_node = NULL;
+
+    irqstate_t flags = enter_critical_section();
+
+    /* Insert pid node in pid list */
+    if (!pid_head) {
+        pid_node = allocate_pid_node(pid);
+        if (!pid_node) {
+            goto exit_with_pidnull;
+        }
+        sq_addfirst((sq_entry_t *)pid_node, &dead_pidList);
+    } else {
+        dead_pid_t *pid_prev;
+        while (pid_head) {
+            if (pid_head->pid == pid) {
```

Comment:
Do we really need to check this? 
Since we will add when task dies and remove when new task is created, I think this check is an overhead. 
If we remove this check, then we can directly insert the new node at the beginning of the list.